File: hearth/llm/gemini.py
"""Token-metered Gemini wrapper. Every Gemini call MUST go through here so
that every call writes a token_usage row — this is the monetization
backbone. No other module may call google-genai directly.
"""
import json
from datetime import datetime, timezone
from typing import Any, Optional

from hearth import config
from hearth.data.store import DataStore


class Gemini:
    def __init__(self, store: DataStore, api_key: Optional[str] = config.GEMINI_API_KEY):
        self.store = store
        self.api_key = api_key
        self._client = None
        if api_key:
            from google import genai

            self._client = genai.Client(api_key=api_key)

    def generate(
        self,
        prompt: str,
        *,
        operation: str,
        user_id: Optional[int] = None,
        request_id: Optional[int] = None,
        system: Optional[str] = None,
        json_schema: Optional[dict] = None,
    ) -> tuple[Any, dict]:
        """Generate text (or parsed JSON if json_schema is given).

        Returns (text_or_parsed_json, usage_dict) and always logs a
        token_usage row, regardless of which branch below runs.
        """
        if self._client is None:
            text, usage = self._stub_generate(prompt, json_schema=json_schema)
        else:
            text, usage = self._real_generate(prompt, system=system, json_schema=json_schema)

        self.store.insert(
            "token_usage",
            {
                "user_id": user_id,
                "operation": operation,
                "model": config.GEMINI_MODEL,
                "prompt_tokens": usage["prompt_tokens"],
                "completion_tokens": usage["completion_tokens"],
                "total_tokens": usage["total_tokens"],
                "request_id": request_id,
                "created_at": datetime.now(timezone.utc).isoformat(),
            },
        )
        return text, usage

    def _real_generate(
        self, prompt: str, *, system: Optional[str], json_schema: Optional[dict]
    ) -> tuple[Any, dict]:
        from google.genai import types

        config_kwargs: dict[str, Any] = {}
        if system:
            config_kwargs["system_instruction"] = system
        if json_schema:
            config_kwargs["response_mime_type"] = "application/json"
            config_kwargs["response_json_schema"] = json_schema

        response = self._client.models.generate_content(
            model=config.GEMINI_MODEL,
            contents=prompt,
            config=types.GenerateContentConfig(**config_kwargs) if config_kwargs else None,
        )

        usage_meta = response.usage_metadata
        usage = {
            "prompt_tokens": usage_meta.prompt_token_count or 0,
            "completion_tokens": usage_meta.candidates_token_count or 0,
            "total_tokens": usage_meta.total_token_count or 0,
        }

        text = response.text
        if json_schema:
            return json.loads(text), usage
        return text, usage

    def _stub_generate(
        self, prompt: str, *, json_schema: Optional[dict]
    ) -> tuple[Any, dict]:
        """Deterministic offline fallback used when GEMINI_API_KEY is unset,
        so the metering path is always testable without network access.
        """
        prompt_tokens = max(1, len(prompt) // 4)
        completion_text = f"[stub response to]: {prompt}"
        completion_tokens = max(1, len(completion_text) // 4)
        usage = {
            "prompt_tokens": prompt_tokens,
            "completion_tokens": completion_tokens,
            "total_tokens": prompt_tokens + completion_tokens,
        }

        if json_schema:
            return self._stub_json(json_schema), usage
        return completion_text, usage

    @staticmethod
    def _stub_json(json_schema: dict) -> dict:
        _STUB_VALUES = {"string": "stub", "number": 0, "integer": 0, "boolean": False}
        result = {}
        for name, prop in json_schema.get("properties", {}).items():
            result[name] = _STUB_VALUES.get(prop.get("type"), None)
        return result

File: src/llm/gemini.ts
import { GoogleGenAI } from "@google/genai";
import { GEMINI_MODEL } from "../config";
import { TokenUsageRow } from "../data/schema";
import { DataStore } from "../data/store";
import { debugLog } from "../debug";
import { GenerateOptions, GenerateResult, UsageMetadata } from "./types";

// Strips ```json fences models sometimes wrap structured output in, then
// re-serializes so callers always get clean, parseable JSON text.
function cleanJsonText(raw: string): string {
  const unfenced = raw.replace(/^```(?:json)?\s*/i, "").replace(/```\s*$/, "");
  try {
    return JSON.stringify(JSON.parse(unfenced));
  } catch {
    return unfenced.trim();
  }
}

// Rough offline heuristic (~4 chars/token) so the stub still produces
// plausible, deterministic-for-the-same-input usage numbers.
function estimateTokens(text: string): number {
  return Math.max(1, Math.ceil(text.length / 4));
}

export default class Gemini {
  private client: GoogleGenAI | null;

  constructor(private store: DataStore, apiKey: string) {
    this.client = apiKey ? new GoogleGenAI({ apiKey }) : null;
  }

  async generate(prompt: string, options: GenerateOptions): Promise<GenerateResult> {
    const { text, usage } = this.client
      ? await this.callGemini(prompt, options)
      : this.stubGenerate(prompt, options);

    await this.store.insert<TokenUsageRow>("token_usage", {
      user_id: options.userId ?? null,
      operation: options.operation,
      model: GEMINI_MODEL,
      prompt_tokens: usage.promptTokens,
      completion_tokens: usage.completionTokens,
      total_tokens: usage.totalTokens,
      request_id: options.requestId ?? null,
      created_at: new Date().toISOString(),
    });

    return { text, usage };
  }

  private async callGemini(prompt: string, options: GenerateOptions): Promise<GenerateResult> {
    debugLog("gemini:request", `operation=${options.operation} grounded=${!!options.grounded}`, {
      model: GEMINI_MODEL,
      prompt,
      system: options.system,
      jsonSchema: options.jsonSchema,
    });

    const response = await this.client!.models.generateContent({
      model: GEMINI_MODEL,
      contents: prompt,
      config: {
        ...(options.system ? { systemInstruction: options.system } : {}),
        ...(options.jsonSchema
          ? { responseMimeType: "application/json", responseSchema: options.jsonSchema }
          : {}),
        ...(options.grounded ? { tools: [{ googleSearch: {} }] } : {}),
      },
    });

    debugLog("gemini:raw-response", `operation=${options.operation}`, response);

    const groundingMetadata = response.candidates?.[0]?.groundingMetadata;
    if (options.grounded && groundingMetadata) {
      debugLog("gemini:websearch", `operation=${options.operation}`, {
        queries: groundingMetadata.webSearchQueries,
        sources: groundingMetadata.groundingChunks?.map((chunk) => chunk.web),
      });
    }

    const rawText = response.text ?? "";
    const text = options.jsonSchema ? cleanJsonText(rawText) : rawText;
    if (text !== rawText) {
      debugLog("gemini:cleaned-text", `operation=${options.operation}`, { rawText, cleanedText: text });
    }

    const usage: UsageMetadata = {
      promptTokens: response.usageMetadata?.promptTokenCount ?? 0,
      completionTokens: response.usageMetadata?.candidatesTokenCount ?? 0,
      totalTokens: response.usageMetadata?.totalTokenCount ?? 0,
    };

    return { text, usage };
  }

  private stubGenerate(prompt: string, options: GenerateOptions): GenerateResult {
    const text = this.stubText(prompt, options);
    debugLog("gemini:stub-response", `operation=${options.operation} (no GEMINI_API_KEY set)`, { prompt, text });

    const promptTokens = estimateTokens(prompt);
    const completionTokens = estimateTokens(text);

    return {
      text,
      usage: {
        promptTokens,
        completionTokens,
        totalTokens: promptTokens + completionTokens,
      },
    };
  }

  private stubText(prompt: string, options: GenerateOptions): string {
    if (options.operation === "request_parse") {
      const request = /raw request:\s*"([^"]+)"/i.exec(prompt)?.[1] ?? prompt;
      const itemMatch = /\bbuy\s+me\s+(.+)$/i.exec(request.trim());
      const itemText = (itemMatch?.[1] ?? request).replace(/[?.!]+$/g, "").trim().toLowerCase();
      const quantityMatch = /^(\d+(?:\.\d+)?\s*(?:x|lb|lbs|oz|pack|packs|bunch|bunches|bag|bags)?)\s+(.+)$/i.exec(
        itemText
      );
      const itemName = (quantityMatch?.[2] ?? itemText).trim();
      const quantity = quantityMatch?.[1] ?? "1";
      return JSON.stringify({ itemName, quantity });
    }

    if (options.operation === "price_sweep") {
      const itemName = /item:\s*([^\n]+)/i.exec(prompt)?.[1]?.trim().toLowerCase() ?? "item";
      const unit = itemName.includes("celery") ? "bunch" : "each";
      const noClear = /ambiguous|tie|no clear|close call/i.test(itemName);
      const candidates = noClear
        ? [
            { store: "Trader Joe's", price: 2.0, unit },
            { store: "Star Market", price: 2.08, unit },
          ]
        : [
            { store: "Trader Joe's", price: 1.99, unit },
            { store: "Star Market", price: 2.49, unit },
            { store: "Whole Foods", price: 2.79, unit },
          ];
      return JSON.stringify({ candidates });
    }

    if (options.operation === "crowd_parse") {
      const reply = /reply:\s*"([^"]+)"/i.exec(prompt)?.[1] ?? prompt;
      const priceMatch = /\$?\s*(\d+(?:\.\d{1,2})?)/.exec(reply);
      const atStoreMatch = /\bat\s+([A-Za-z0-9 '&.-]+)/i.exec(reply);
      const storeMatch = /^([A-Za-z0-9 '&.-]+?)\s+(?:has|is|for|\$)/i.exec(reply);
      return JSON.stringify({
        store: (atStoreMatch?.[1] ?? storeMatch?.[1] ?? "Unknown store").trim(),
        price: priceMatch ? Number(priceMatch[1]) : null,
      });
    }

    if (options.operation === "growth_post") {
      const stat = /stat:\s*([\s\S]+)/i.exec(prompt)?.[1]?.trim() ?? prompt.trim();
      return `Hearth price watch: ${stat}`;
    }

    if (options.jsonSchema) {
      return JSON.stringify({ stub: true, operation: options.operation });
    }

    return `[stub:${GEMINI_MODEL}] ${prompt}`;
  }
}
